canister/src/unstable_blocks.rs

```diff
@@ -244,43 +252,160 @@ mod test {
         assert_eq!(peek(&forest), Some(&block_0));
         assert_eq!(pop(&mut forest), Some(block_0));
 
-        // Block 1 is now the anchor. It doesn't have children yet,
-        // so calling `pop` should return `None`.
+        // Block 1 is now the anchor. It doesn't have stable
+        // children yet, so calling `pop` should return `None`.
+        assert_eq!(peek(&forest), None);
+        assert_eq!(pop(&mut forest), None);
+    }
+
+    #[test]
+    fn single_chain_various_difficulties() {
+        let block_0 = BlockBuilder::genesis().build().with_mock_dificulty(20);
+        let block_1 = BlockBuilder::with_prev_header(block_0.header())
+            .build()
+            .with_mock_dificulty(15);
+        let block_2 = BlockBuilder::with_prev_header(block_1.header())
+            .build()
+            .with_mock_dificulty(20);
+        let block_3 = BlockBuilder::with_prev_header(block_2.header())
+            .build()
+            .with_mock_dificulty(110);
+
+        let utxos = UtxoSet::new(Network::Mainnet);
+        let mut forest = UnstableBlocks::new(&utxos, 7, block_0.clone());
+
+        push(&mut forest, &utxos, block_1.clone()).unwrap();
+        push(&mut forest, &utxos, block_2).unwrap();
+        assert_eq!(peek(&forest), None);
+        assert_eq!(pop(&mut forest), None);
+
+        push(&mut forest, &utxos, block_3).unwrap();
+        // Block 0 (the anchor) now has one stable child (Block 1).
+        // Block 0 should be returned when calling `pop`.
+        assert_eq!(peek(&forest), Some(&block_0));
```

Comment:
nit: I think it would be useful here and elsewhere in tests to explain why block 1 has now become stable. It's as obvious anymore with the additional complexity we're introducing with `difficulty`.